cornac/models/tifuknn/recom_tifuknn.py

```diff
@@ -0,0 +1,171 @@
+        The name of the recommender model.
+
+    n_neighbors: int, optional, default: 300
+        The number of neighbors for KNN
+
+    within_decay_rate: float, optional, default: 0.9
+        Within-basket time-decayed ratio in range [0, 1]
+
+    group_decay_rate: float, optional, default: 0.7
+        Group time-decayed ratio in range [0, 1]
+
+    alpha: float, optional, default: 0.7
+        The trade-off between current user vector and neighbors vectors
+        to compute final item scores
+
+    n_groups: int, optional, default: 7
+        The historal baskets will be partition into `n_groups` equally.
+
+    verbose: boolean, optional, default: False
+        When True, running logs are displayed.
+
+    References
+    ----------
+    Haoji Hu, Xiangnan He, Jinyang Gao, and Zhi-Li Zhang. 2020.
+    Modeling Personalized Item Frequency Information for Next-basket Recommendation.
+    In Proceedings of the 43rd International ACM SIGIR Conference on Research and Development in Information Retrieval (SIGIR '20). Association for Computing Machinery, New York, NY, USA, 1071–1080. https://doi.org/10.1145/3397271.3401066
+
+    """
+
+    def __init__(
+        self,
+        name="TIFUKNN",
+        n_neighbors=300,
+        within_decay_rate=0.9,
+        group_decay_rate=0.7,
+        alpha=0.7,
+        n_groups=7,
+        verbose=False,
+    ):
+        super().__init__(name=name, trainable=False, verbose=verbose)
+        assert within_decay_rate >= 0 and within_decay_rate <= 1
+        assert group_decay_rate >= 0 and group_decay_rate <= 1
+        self.n_neighbors = n_neighbors
+        self.within_decay_rate = within_decay_rate
+        self.group_decay_rate = group_decay_rate
+        self.alpha = alpha
+        self.n_groups = n_groups
+
+    def fit(self, train_set, val_set=None):
+        from scipy.spatial import KDTree
+
+        super().fit(train_set=train_set, val_set=val_set)
+        self.user_vectors = self._get_user_vectors(self.train_set)
+        if self.n_neighbors > len(self.user_vectors):
+            warnings.warn("Number of users is %d, smaller than number of neighbors %d" % (len(self.user_vectors), self.n_neighbors))
+            self.n_neighbors = len(self.user_vectors)
+
+        start_time = time()
+        if self.verbose:
+            print("Constructing kd-tree for quick nearest-neighbor lookup")
+        self.tree = KDTree(list(self.user_vectors.values()))
+        if self.verbose:
+            print("Constructing kd-tree for quick nearest-neighbor lookup takes %.0f" % (time() - start_time))
+        return self
+
+    def _get_user_vectors(self, data_set):
+        user_vectors = {}
+        for [user_idx], _, [basket_items] in tqdm(
+            data_set.ubi_iter(batch_size=1, shuffle=False),
+            desc="Getting user vectors",
+            total=data_set.num_users,
+        ):
+            user_vectors[user_idx] = self._compute_user_vector(basket_items[:-1])
```

Comment:
do we need `user_vectors` to be a dic? Is it possible for it to be a numpy 2D array?